src/main/java/pl/edu/agh/firecell/engine/algorithm/SmokePropagator.java

```diff
@@ -0,0 +1,129 @@
+        if (oldState.hasCell(NeighbourUtils.down(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.down(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.down(cellIndex)).isSolid()) {
+            smokeFromFire = oldState.getCell(NeighbourUtils.down(cellIndex)).material().smokeCoe();
+        }
+        Optional<Cell> northCell = Optional.empty();
+        Optional<Cell> southCell = Optional.empty();
+        Optional<Cell> westCell = Optional.empty();
+        Optional<Cell> eastCell = Optional.empty();
+        if (oldState.hasCell(NeighbourUtils.west(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.west(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.west(cellIndex)).isSolid()) {
+            westCell = Optional.of(oldState.getCell(NeighbourUtils.west(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.east(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.east(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.east(cellIndex)).isSolid()) {
+            eastCell = Optional.of(oldState.getCell(NeighbourUtils.east(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.north(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.north(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.north(cellIndex)).isSolid()) {
+            northCell = Optional.of(oldState.getCell(NeighbourUtils.north(cellIndex)));
+        }
+        if (oldState.hasCell(NeighbourUtils.south(cellIndex)) &&
+                isCellBurning(oldState.getCell(NeighbourUtils.south(cellIndex))) &&
+                oldState.getCell(NeighbourUtils.south(cellIndex)).isSolid()) {
+            southCell = Optional.of(oldState.getCell(NeighbourUtils.south(cellIndex)));
+        }
+        if (southCell.isPresent()) smokeFromFire += southCell.get().material().smokeCoe() / 4;
+        if (northCell.isPresent()) smokeFromFire += northCell.get().material().smokeCoe() / 4;
+        if (westCell.isPresent()) smokeFromFire += westCell.get().material().smokeCoe() / 4;
+        if (eastCell.isPresent()) smokeFromFire += eastCell.get().material().smokeCoe() / 4;
+
+        double smokeDifference = getSmokeIndicatorDifference(oldCell, oldState, cellIndex);
+        if(Math.min(oldCell.smokeIndicator() + smokeDifference + smokeFromFire, 100) < 0.0005)
+            return 0.0;
+        return Math.min(oldCell.smokeIndicator() + smokeDifference + smokeFromFire, 100);
+    }
+
+    private double getSmokeIndicatorDifference(Cell oldCell, State oldState, Vector3i cellIndex) {
+        Int counter = new Int(oldCell.smokeIndicator());
+        return NeighbourUtils.neighboursStream(cellIndex)
+                .filter(oldState::hasCell)
+                .filter(neighbourIndex -> oldState.getCell(neighbourIndex).isFluid())
+                .map(neighbourIndex -> {
+                    double diff = 0.0;
```

Comment:
Delete this variable ...

Reply:
Why? Ok i can return the value in first and second if statement but in third? And in first two, I need to update current value and return it, still extra variable.